src/AssetControlExtension.php

```diff
@@ -209,6 +217,18 @@ protected function addAssetsFromRecord(AssetManipulationList $manipulation, Data
         }
     }
 
+    private function hasAssets(): bool
+    {
+        $fields = DataObject::getSchema()->fieldSpecs($this->owner);
+        foreach ($fields as $field => $db) {
+            $fieldObj = $this->owner->$field;
+            if ($fieldObj instanceof DBFile) {
+                return true;
+            }
+        }
```

Comment:
@GuySartorelli i may be wrong here, but as far as i understand, fieldSpecs do not take into consideration versions, so checking the schema should return the same thing regardless

if we want to be 100% sure, we can also simply only wrap this part in my new hasAssets method

        // Mark overwritten object as deleted
        if ($this->owner->isInDB()) {
            $priorRecord = DataObject::get(get_class($this->owner))->byID($this->owner->ID);
            if ($priorRecord) {
                $this->addAssetsFromRecord($manipulations, $priorRecord, AssetManipulationList::STATE_DELETED);
            }
        }

this way, we get rid of the priorRecord call, which is the main issue, and add a todo for later to optimize things even further

Reply:
`fieldSpecs()` will always return the same thing, but we're not checking against that, we're checking if `$this->owner->$field` is an instance of `DBFile` - in other words does the version of the record we currently have in memory have a `DBFile` as the _value_ for that field. It's checking what's stored in the db for a given version of a record, not just checking directly against the static schema.

If something like this works, it would make me a lot more at ease about it:

```php
private function hasAssets(): bool
{
    $fields = DataObject::getSchema()->fieldSpecs($this->owner);
    foreach ($fields as $field => $fieldSpec) {
        $fieldClass = strtok($fieldSpec ?? '', '(');
        if (singleton($fieldClass) instanceof DBFile) {
            return true;
        }
    }
    return false;
}
```

The above is based mostly from memory, so may need some tweaking - but basically instead of checking the value of the field on the class, just check purely what's in the field specs.

However, while writing this I realised another potential regression this could cause - definitely an edge case, but something to consider nevertheless. If someone puts a `DBFile` field wrapped inside a `DBComposite` for some reason (not likely but definitely not impossible), this will always say there's no assets, even when there is.
I'm not sure what the current behaviour is in that scenario though, so maybe that already doesn't work?

Reply:
@lekoala just `@`ing you here in case you missed this comment.

Reply:
@GuySartorelli i see your point. Using singleton creates other issues (here, it will fail on DBPrimaryKey because it has no arguments). I still think it's fine to use the current instance as a reference. dbObject will return an empty object anyway based on the schema specs, so this should end up being the same, with little to no extra work involved. i've changed my code to call dbObject directly to make sure we do get a proper class based on the specs

Reply:
> dbObject will return an empty object anyway based on the schema specs

It won't necessarily be an _empty_ object - if the field has a value it just returns the field including the value. But it does importantly return a spec-based object regardless of whether the record we're calling it on has a value so I'm okay with this. Thank you for making the change.